File: pack/logicmoo_base/prolog/logicmoo/planner/frolog/Main/frolog.java
package Main;


import UserInterface.PlayerInterface;
import jpl.*;
import de.tuebingen.ui.CommandLineOptions;
import de.tuebingen.tree.Grammar;
import NLPModules.Parsing.*;
import java.util.*; 



/**
 * This is the entry point of the application
 *  
 * There are several libraries that need to be loaded: gecodej, jpl
 * I've specified this as an argument to the VM in the eclipse project like this:
 * -Djava.library.path="/usr/lib/pl-5.6.54/lib/i386-linux:/usr/local/lib"
 * May be it's better to load them in the code using System.load(arg0)
 * 
 * @author benottil
 *
 */

public class frolog {

	/**
	 * @param args
	 */
	static Process geni;
	static Process racer;
	public static CommandLineOptions tulipaOptions;
	public static Grammar tulipaGrammar;
//	public static String[] thematicRoles= {"agent","patient","instrument","source","goal","instrument"};
	public static Set<String> discourseModel = new HashSet<String>();
	public static String lastIndividual = ""; 
	public static String PreferedOrdering = "['colour','(some haslocation *top*)','(some hasdetail *top*)']";
	//primitive concepts are those concepts that are not defined in the tbox, 
	//they are either in the game abox or asserted through action effects
	public static List<String> primitiveConcepts = Arrays.asList("drawingroom","treasury","room","wall","green","white",
		"player","alive","frog","ugly","brown","sword","small", "crown","silver","apple","dragon","worm",
		"red","chest","wooden","locked","unlocked","closed","open","couch","couchleg","golden","key","table","southexit","northexit",
		"hasdetail","hold","fitsin","hasexit","leadsto","hascounterpart","gone","happy","beautiful","pizza","yellow");
	
	
	public static void main(String[] args) {
		
//		System.load("/usr/lib/pl-5.6.54/lib/i386-linux/libjpl.so");
		
		initialize();
		
		PlayerInterface.startInterface();
		
		try {
			geni.destroy();
			System.out.println("Geni closed properly");
			racer.destroy();
			System.out.println("Racer closed properly");
		
		} catch (Exception e) {
    		System.err.println(e);
    		System.exit(0); 
		}
	}
	
	public static void initialize(){
	
		
		try {
			//I am not using morphology in geni because it's not working (see geni mailing list) --morphlexicon GameScenarios/FairyTaleCastle/ObjectGrammarGeni/morph.mph
			//Alexandre Denis has this option when calling Geni, I don't know what's its purpose --rootfeat='[cat:Nom|NomPropre|Sen|Prix|Pro]'
			geni = Runtime.getRuntime().exec("NLPModules/Realization/geniserver -m GameScenarios/FairyTaleCastle/ObjectGrammarGeni/lu-small-grammar.geni -l GameScenarios/FairyTaleCastle/ObjectGrammarGeni/lexicon.geni --rootfeat='[cat:s|np]' &");
			System.out.println("Geni (my generator) started successfully");
			
			//racer = Runtime.getRuntime().exec("KBInterfaces/RacerInterface/RacerPro-1-9-2-beta/RacerPro -silent > racer_log.txt &");
			racer = Runtime.getRuntime().exec("KBInterfaces/RacerInterface/RacerPro-Linux32-1-9-3-Beta/RacerPro -silent &");
			System.out.println("Racer (my knowledge base manager) started successfully");
			
			new Query("consult('KBInterfaces/RacerInterface/racer.pl')").hasSolution();			 
			Query q3 = new Query("racer:load_scenario(f)");
			System.out.println(q3.hasSolution() ? "The game scenario knowledge bases were loaded" : "The game scenario failed to load");
		
			String[] commandline = "-g GameScenarios/FairyTaleCastle/ObjectGrammarTulipa/lu-small-grammar.xml -l GameScenarios/FairyTaleCastle/ObjectGrammarTulipa/lexicon.xml -m GameScenarios/FairyTaleCastle/ObjectGrammarTulipa/morph.xml -a s -i -n".split(" ");
			tulipaOptions = Tulipa.processCommandLine(commandline);
			tulipaGrammar = Tulipa.initializeTulipa(tulipaOptions);
			System.out.println("Tulipa (my parser) started successfully");
			
			new Query("consult('NLPModules/ReferenceGeneration/referenceGeneration.pl')").hasSolution();
			new Query("consult('NLPModules/ReferenceResolution/resolve.pl')").hasSolution();
			new Query("consult('NLPModules/Actions/execute.pl')").hasSolution();
			new Query("consult('NLPModules/ContentDetermination/describe.pl')").hasSolution();
			new Query("consult('GameScenarios/FairyTaleCastle/actionDatabase.pl')").hasSolution();
			
			
			} catch (Exception e) {
				System.err.println(e);
			}	
	}
} 

File: pack/logicmoo_base/prolog/logicmoo/planner/frolog/frolog.pl
/*************************************************************************

    File: frolog.pl
    Copyright (C) 2007 

    Programmer: Luciana Benotti

    This file is part of Frolog, version 0.1 (October 2007).

    Frolog is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Frolog is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Frolog; if not, write to the Free Software Foundation, Inc., 
    59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*************************************************************************/

:- module(frolog,[load/1,execute/2, resolveReso/2]).  

:- use_module('KBInterfaces/RacerInterface/racer',[load_scenario/5]).
:- use_module('NLPModules/Actions/execute',[executeCommand/4]).
:- use_module('NLPModules/ContentDetermination/describe',[describe/2,describe/1]).
:- use_module('NLPModules/ReferenceResolution/resolve',[resolveResolutionMain/3]).
:- use_module('GameScenarios/FairyTaleCastle/actionDatabase').

% Don't load the scenario right away after starting racer because racer takes a while to start. 
startRacer :- shell('/users/led/benottil/Sync/Programming/Racer/RacerPro/RacerPro-1-9-2-beta/RacerPro -silent > racer_log.txt &').


load(Scenario) :- 
	retractall(racer:worldKB(_,_)), 
	retractall(racer:playerKB(_,_)),
	load_scenario(Scenario,_,TBoxName,GameAboxName,PlayerAboxName),
	format('The Name of the TBox is: ~w ~n',TBoxName), 
	format('The Name of the Game ABox is: ~w ~n',GameAboxName), 
	format('The Name of the Player ABox is: ~w ~n',PlayerAboxName).

end :- shell('killall RacerPro').

execute(Command,Executed,Failed,Decision) :- executeCommand(Command,Executed,Failed,Decision).

resolveReso(Concept, I):- resolveResolutionMain(Concept,I,DM).



	

